src/Php85/Php85.php

```diff
@@ -33,4 +33,21 @@ public static function get_exception_handler(): ?callable
 
         return $handler;
     }
+
+    public static function array_first(array $array)
+    {
+        foreach ($array as $value) {
+            return $value;
+        }
+        return null;
+    }
+
+    public static function array_last(array $array)
+    {
+        if ($array === []) {
+            return null;
+        }
+
+        return end($array);
```

Comment:
I'm just wondering if using end() triggers copy-on-write?
```suggestion
        return $array ? end($array) : null;
```

Reply:
> I'm just wondering if using end() triggers copy-on-write?

Yes.

```
<?php

function array_last(array $array)
{
	if ($array === []) {
		return null;
	}

	echo "before end: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;
	$value = end($array);
	echo "after end: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;

	return $value;
}
function array_last_slice(array $array)
{
	if ($array === []) {
		return null;
	}

	echo "before slice: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;
	[$value] = array_slice($array, -1, 1);
	echo "after slice: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;

	return $value;
}

echo "initial: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;

$a = array_fill(0, 100_000, new stdClass());

echo "after fill: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;

var_dump(array_last($a));

echo "after last: ", memory_get_peak_usage() / 1024 / 1024, PHP_EOL;
```

Outputs for me:

```
initial: 0.41307830810547
after fill: 2.3816223144531
before end: 2.3816604614258
after end: 4.3856353759766
object(stdClass)#1 (0) {
}
after last: 4.3856735229492
```

for the `end()` implementation, doubling the array memory and:

```
initial: 0.4130859375
after fill: 2.3816299438477
before slice: 2.3816680908203
after slice: 2.3818359375
object(stdClass)#1 (0) {
}
after last: 2.3819274902344
```

for the `array_slice()` implementation, which is barely noticeable.

Reply:
Wow it really does double the memory footprint! I rewrote with an array_slice call, but without preserving the keys and directly accessing the `0`th key.